Agents/BMSUpdateAgent/BMSUpdateAgent/src/main/java/uk/ac/cam/cares/jps/agent/bmsupdate/BMSUpdateAgentLauncher.java

```diff
@@ -107,66 +156,163 @@ public JSONObject processRequestParameters(JSONObject requestParams) {
+                // pass data IRI, remote store client, remote RDB client. triggerValue, insertString, deleteString to checkInsertAndDelete method
+                subMessage = bmsUpdateAgent.checkInsertAndDelete(dataIRI, rsClient, RDBClient, triggerValue, insertString, deleteString);
+                result.put("message" + i, subMessage);
+            } else if (checkAndUpdateParameters.has(KEY_INSERT) && !checkAndUpdateParameters.has(KEY_DELETE)) {
+                insertString = checkAndUpdateParameters.getString(KEY_INSERT);
+                subMessage = bmsUpdateAgent.checkInsert(dataIRI, rsClient, RDBClient, triggerValue, insertString);
+                result.put("message" + i, subMessage);
+                // pass data IRI, remote store client, remote RDB client. triggerValue, insertString to checkInsert method
+            } else if (checkAndUpdateParameters.has(KEY_DELETE) && !checkAndUpdateParameters.has(KEY_INSERT)) {
+                deleteString = checkAndUpdateParameters.getString(KEY_DELETE);
+                // pass data IRI, remote store client, remote RDB client. triggerValue, deleteString to checkDelete method
+                subMessage = bmsUpdateAgent.checkDelete(dataIRI, rsClient, RDBClient, triggerValue, deleteString);
+                result.put("message" + i, subMessage);
+            }
         }
+        return result;
+    }
 
-        String clientProperties = requestParams.getString(KEY_CLIENT_PROPERTIES);
-        if (clientProperties.equals("CLIENT_PROPERTIES")) {
-            if (System.getenv(clientProperties) == null) {
-                LOGGER.error("CLIENT_PROPERTIES not set in environment variable, expect a valid file path");
-                return false;
+    /**
+     * execute update present value route
+     * @param requestParams Request parameters parsed from the body of POST request
+     * @return results of executed route
+     */
+    public JSONObject executeUpdatePresentValue(JSONObject requestParams) {
+        JSONObject result = new JSONObject();
+        JSONObject subMessage = new JSONObject();
+        for ( int i = 0; i < requestParams.getJSONArray(KEY_DATAIRI).length(); i++) {
+                String dataIRI = requestParams.getJSONArray(KEY_DATAIRI).getString(i);
+                String clientPropertiesFile = System.getenv(requestParams.getString(KEY_CLIENT_PROPERTIES));
+                BMSUpdateAgent bmsUpdateAgent = new BMSUpdateAgent();
+                try {
+                    initSparqlProperties(clientPropertiesFile);
+                } catch (IOException e) {
+                    throw new JPSRuntimeException("Unable to read the client properties file.", e);
+                }
+                rsClient = new RemoteStoreClient(sparqlQueryEndpoint, sparqlUpdateEndpoint);
+                if (sparqlUsername != null && sparqlPassword != null) {
+                    rsClient.setUser(sparqlUsername);
+                    rsClient.setPassword(sparqlPassword);
+                }
```

Comment:
Should move line 279-290 out of the for loop. Since the client for each check is the same, will be faster if share the same agent and rsClient instance.

Reply:
Refactored executeUpdatePresentValue to be similar to executeUpdateTriples. Users are now allowed to input different client.properties files for each check. The users just have to create a properties file with the appropriate keys and values and update the "ENV" section in the stack json file before spinning up the agent.